src/opencl/non_nvidia.cl

```diff
@@ -0,0 +1,489 @@
+    uchar i;
+
+    for (mask = 0, i = 0; i < sizeof(vec256) / sizeof(limb_t); i++)
+        mask |= a[i];
+
+    mask = cbit & (is_zero(mask) ^ 1);
+    mask = 0 - mask;
+
+    out[0] = a[0] + (189 & mask); carry = out[0] < (189 & mask);
+    for (i = 1; i < sizeof(vec256) / sizeof(limb_t); i++)
+        out[i] = a[i] + carry, carry = out[i] < carry;
+    for (i = 0; i < sizeof(vec256) / sizeof(limb_t); i++)
+        out[i] ^= mask;
+    out[0] += (1 & mask); carry = out[0] < (1 & mask);
+    for (i = 1; i < sizeof(vec256) / sizeof(limb_t); i++)
+        out[i] += carry, carry = out[i] < carry;
+}
+
+static bool xor_n_check_mod_256_189(vec256 out, const vec256 a, const vec256 b)
+{
+    uchar i;
+    bool carry;
+
+    for (i = 0; i < sizeof(vec256) / sizeof(limb_t); i++)
+        out[i] = a[i] ^ b[i];
+
+    carry = (out[0] + 189) < 189;
+    for (i = 1; i < sizeof(vec256) / sizeof(limb_t); i++)
+        carry = (out[i] + carry) < carry;
+
+    return carry;
+}
+
+typedef uint u32;
+typedef ulong u64;
+typedef u32 fe26[10];           /* 9x26+22=256 */
+
+#define MASK26 0x3ffffff
+#define MASK22 0x3fffff
+
+static void to_fe26(fe26 out, const void* in_)
+{
+    const u32* in = (const u32*)in_;
+
+    out[0] = (in[0]) & MASK26;
+    out[1] = (in[0] >> 26 | in[1] << 6) & MASK26;
+    out[2] = (in[1] >> 20 | in[2] << 12) & MASK26;
+    out[3] = (in[2] >> 14 | in[3] << 18) & MASK26;
+    out[4] = (in[3] >> 8 | in[4 ] << 24) & MASK26;
+    out[5] = (in[4] >> 2) & MASK26;
+    out[6] = (in[4] >> 28 | in[5] << 4) & MASK26;
+    out[7] = (in[5] >> 22 | in[6] << 10) & MASK26;
+    out[8] = (in[6] >> 16 | in[7] << 16) & MASK26;
+    out[9] = (in[7] >> 10);
+}
+
+static void from_fe26(void* out_, const fe26 in)
+{
+    u32 h0, q, * out = (u32*)out_;
+    u64 t;
+
+    /* compare to the modulus */
+    q = ((h0 = in[0]) + 189) >> 26;
+    q = (in[1] + q) >> 26;
+    q = (in[2] + q) >> 26;
+    q = (in[3] + q) >> 26;
+    q = (in[4] + q) >> 26;
+    q = (in[5] + q) >> 26;
+    q = (in[6] + q) >> 26;
+    q = (in[7] + q) >> 26;
+    q = (in[8] + q) >> 26;
+    q = (in[9] + q) >> 22;
+
+    /* full reduction */
+    h0 += 189 * q;
+
+    t = (u64)h0;
+    t += (u64)in[1] << 26;    out[0] = (u32)t;    t >>= 32;
+    t += (u64)in[2] << 20;    out[1] = (u32)t;    t >>= 32;
+    t += (u64)in[3] << 14;    out[2] = (u32)t;    t >>= 32;
+    t += (u64)in[4] << 8;     out[3] = (u32)t;    t >>= 32;
+    t += (u64)in[5] << 2;
+    t += (u64)in[6] << 28;    out[4] = (u32)t;    t >>= 32;
+    t += (u64)in[7] << 22;    out[5] = (u32)t;    t >>= 32;
+    t += (u64)in[8] << 16;    out[6] = (u32)t;    t >>= 32;
+    t += (u64)in[9] << 10;    out[7] = (u32)t;
+}
+
+#if defined(__INTEL_GPU__)
+inline u64 mul(u32 x, u32 y)
+{
+    return (u64)x * y;
+}
+
+inline u64 mad(u64 z, u32 x, u32 y)
+{
+    return (u64)x * y + z;
+}
+#elif defined(__AMD_GPU__)
+inline u64 mul(u32 x, u32 y)
+{
+    union { u64 W; u32 w[2]; } z;
+    __asm__("v_mul_lo_u32 %0, %1, %2"

```

Comment:
@simonatsn we have faced an issue on Windows where this may not compile, see https://github.com/subspace/sloth256-189/issues/29

I'm wondering if these ASM parts are mandatory or we can fall back to the same code as Intel GPUs use above if `__asm__` macro is not available?

Reply:
Mostly likely the code for the Intel GPU won't work. In the past we had trouble with the AMD compiler producing incorrect results for that code, possibly due to incorrect optimizations. 

However I expect that code should work on Windows. Is it possible the version of opencl being used is old? Can you obtain the results of clinfo? It should be using opencl2 or newer.

Reply:
I think that makes sense (filtering out particular GPUs if they are problematic), especially for the smaller GPUs that don't contribute much to overall performance. 

Another thing to consider adding is a version check to make sure the OpenCL driver is recent. You should be able to call clGetPlatformInfo to get the CL_PLATFORM_VERSION param and make sure it's >= 2.

Reply:
Did that in https://github.com/subspace/sloth256-189/pull/30, thanks for suggestion!